app/App/Main/Account/Requests/TransactionRequest/AdjustFee/index.js

```diff
@@ -25,17 +45,156 @@ function maxFee (tx = { chainId: '' }) {
+    const limitedValue = limiter(newValue)
+    onReceiveValue(limitedValue)
+    setValue(formatForInput(limitedValue, decimals))
+  }
+  const set = (newValueStr, newValueNum) => {
+    setValue(newValueStr)
+    clearTimeout(submitTimeout)
+
+    setSubmitTimeout(
+      setTimeout(() => processValue(newValueNum), 500)
+    )
+  }
+
+  return (
+    <>
+      <div className='txFeeOverlayInput'>
+        <input 
+          tabIndex={tabIndex} 
+          value={value}
+          className='txFeeOverlayInput' 
+          onChange={(e) => {
+            const value = (decimals ? /[0-9\.]*/ : /[0-9]*/).exec(e.target.value)
+            if (value) {
+              // special case to prevent decimal point being overwritten as user is typing a float
+              if (value[0].endsWith('.')) {
+                const formattedNum = formatForInput(value[0].slice(0, -1), decimals)
+                setValue(`${formattedNum}.`)
+                clearTimeout(submitTimeout)
+                return
+              }
+
+              const parsedValue = decimals ? parseFloat(value[0]) : parseInt(value[0])
+              if (isNaN(parsedValue)) {
+                return
+              }
+
+              set(value[0], parsedValue)
+            }
+          }}
+          onKeyDown={(e) => {
+            if (e.key === 'Enter') {
+              e.preventDefault()
+              e.target.blur()
+            } else if (e.key === 'ArrowUp' || e.key === 'ArrowDown') {
+              e.preventDefault()
+              const parsedValue = decimals ? parseFloat(value) : parseInt(value)
+              if (isNaN(parsedValue)) {
+                return
+              }
+
+              if (e.key === 'ArrowUp') {
+                const incrementedValue = decimals ? limitRange(Math.floor(parsedValue) + 1) : (parsedValue + 1000)
+                set(incrementedValue)
+              } else {
+                const decrementedValue = decimals ? limitRange(Math.floor(parsedValue) - 1) : (parsedValue - 1000)
+                set(decrementedValue)
+              }
+            }
+          }}
+        />
+        </div>
+        <div className='txFeeOverlayLabel'>{labelText}</div>
+      </>
+    )
+}
+
+const GasLimitInput = ({ initialValue, maxTotalFee, baseFee, priorityFee, gasPrice, handlerId }) => {
+  const receiveValueHandler = (newValue) => {
+    let gasLimit = newValue
+    if (gasPrice && gweiToWei(gasPrice) * gasLimit > maxTotalFee) {
+      gasLimit = Math.floor(maxTotalFee / gweiToWei(gasPrice))
+    } else if (gweiToWei(baseFee + priorityFee) * newValue > maxTotalFee) {
+      gasLimit = Math.floor(maxTotalFee / gweiToWei(baseFee + priorityFee))
+    }
+
+    link.rpc('setGasLimit', intToHex(gasLimit), handlerId, (e) => {
+      if (e) console.error(e)
+    })
+  }
+
+  return (
+    <div className='txFeeOverlayLimit'>
+      <FeeOverlayInput initialValue={initialValue} onReceiveValue={receiveValueHandler} labelText='Gas Limit (UNITS)' tabIndex={2} decimals={false} />
```

Comment:
I think it might make more sense to pass `tabIndex` down from the parent component that way if we re-arrange them this won't break

Reply:
You mean build an array of all the inputs then loop through them and pass the index as tabIndex?

Reply:
this is probably ideal, but we could also just hardcode them when we create the components. either way they should reflect the order they're laid out on the screen